flink-ml-lib/src/main/java/org/apache/flink/ml/feature/standardscaler/OnlineStandardScalerModel.java

```diff
@@ -0,0 +1,298 @@
+
+        /** Inverse of standard deviation. */
+        private DenseVector scale;
+
+        private long modelVersion;
+
+        private long modelTimeStamp;
+
+        public PredictionOperator(
+                RowTypeInfo inputTypeInfo,
+                String inputCol,
+                boolean withMean,
+                boolean withStd,
+                long maxAllowedModelDelayMs,
+                String modelVersionCol) {
+            this.inputTypeInfo = inputTypeInfo;
+            this.inputCol = inputCol;
+            this.withMean = withMean;
+            this.withStd = withStd;
+            this.maxAllowedModelDelayMs = maxAllowedModelDelayMs;
+            this.modelVersionCol = modelVersionCol;
+        }
+
+        @Override
+        public void initializeState(StateInitializationContext context) throws Exception {
+            super.initializeState(context);
+            bufferedPointsState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<StreamRecord>(
+                                            "bufferedPoints",
+                                            new StreamElementSerializer(
+                                                    inputTypeInfo.createSerializer(
+                                                            getExecutionConfig()))));
+
+            modelDataState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>(
+                                            "modelData",
+                                            TypeInformation.of(StandardScalerModelData.class)));
+        }
+
+        @Override
+        public void snapshotState(StateSnapshotContext context) throws Exception {
+            super.snapshotState(context);
+            if (modelData != null) {
+                modelDataState.clear();
+                modelDataState.add(modelData);
+            }
+        }
+
+        @Override
+        public void open() throws Exception {
+            super.open();
+            MetricGroup mlModelMetricGroup =
+                    getRuntimeContext()
+                            .getMetricGroup()
+                            .addGroup(MLMetrics.ML_GROUP)
+                            .addGroup(MLMetrics.ML_MODEL_GROUP);
+            mlModelMetricGroup.gauge(MLMetrics.TIMESTAMP, (Gauge<Long>) () -> modelTimeStamp);
+            mlModelMetricGroup.gauge(MLMetrics.VERSION, (Gauge<Long>) () -> modelVersion);
+        }
+
+        @Override
+        public void processElement1(StreamRecord<Row> dataPoint) throws Exception {
+            if (dataPoint.getTimestamp() <= modelTimeStamp + maxAllowedModelDelayMs
+                    && mean != null) {
+                doPrediction(dataPoint);
+            } else {
+                bufferedPointsState.add(dataPoint);
```

Comment:
Given that we only update `modelDataState` in `snapshotState()`, I suppose it is because updating `modelDataState()` every time we receive a model data is more expensive than updating the in-memory `modelData`.

Should we follow the same approach for `dataPoint` consistency and performance?

Reply:
There is a bit difference between `modelDataState` and `bufferedPointsState` --- There is one single element in model data in `modelDataState` and more than one elements in `bufferedPointsState`.

Appending elements into liststate is an asyncronous operation and thus can help to improve the efficiency when caching data points. On the other side, there is only one model data and we updating it may introduce unnessary state  operations (e.g., add it to state and remove it).

Reply:
Thanks for the explanation. Sounds good.